### PR TITLE
cli: modify job-trace to collect cluster wide spans

### PR DESCRIPTION
**Pull request by adityamaru**
Previously, `debug job-trace` would contact the local
inflight trace registry, and dump the StrcuturedRecords
for each span in a csv file. This was not very
consumable when debugging using traces AND we now have
a way to collect inflight traces for a given TraceID from
all nodes in the cluster.

This change does a few things:

- Instead of contacting the inflight node registry, we now
contact the indexed vtable crdb_internal.cluster_inflight_traces
that returns a row per `tracing.Recording` for every node with
inflight spans with the given TraceID.

- For each node we accumulate three formats for all `tracing.Recordings`
in a node, and flush this information to three files:
`nodeID-trace.json`, `nodeID-trace.txt`, and `nodeID-jaeger.json`

For the JSON and txt formats, each tracing.Recording within a node
is separated by the root span operation name.

For the JaegerJSON we "stitch" the jaeger formats for all
`tracing.Recordings` in a node. Refer to method comments to
see how we do this, and why it is okay to do.

Release note (cli change): `cockroach debug job-trace` now
creates a `job-trace.zip` which contains trace information
for each node executing the job.

### CODE REVIEW

**Comment by adityamaru** [comment]
The first commit is from https://github.com/cockroachdb/cockroach/pull/66679, will rebase once that is merged!

**Comment by adityamaru** [comment]
I might remove the raw JSON based on feedback from a slack thread of it not being a useful format, but otherwise, this should be ready for a review. 

For reference this is the output zip of a "stuck" backup job: 
[job-trace.zip](https://github.com/cockroachdb/cockroach/files/6729493/job-trace.zip)

**Comment by adityamaru** [comment]
cc: @abarganier since you've been the primary reviewer on most of the underlying code driving this.

File: pkg/cli/debug_job_trace_test.go
Comment (pbardea) [inline pkg/cli/debug_job_trace_test.go:126]: It looks like this supports being called as `checkBundle(t, id, "node1-trace.txt", "node1-trace.json", "node1-jaeger.json")` -- was that the intention?

File: pkg/cli/debug_job_trace.go
Comment (pbardea) [inline pkg/cli/debug_job_trace.go:125]: nit: perhaps simpler:
```
cumulativeTraceCollection.Data[0].Spans = append(cumulativeTraceCollection.Data[0].Spans, curTraceCollection.Data[0].Spans...)
```

File: pkg/cli/debug_job_trace.go
Comment (pbardea) [inline pkg/cli/debug_job_trace.go:112]: Can we move this case further up the method to be the first check?

File: pkg/cli/debug_job_trace.go
Comment (pbardea) [inline pkg/cli/debug_job_trace.go:137]: nit: extra new-line

File: pkg/cli/debug_job_trace.go
Comment (pbardea) [inline pkg/cli/debug_job_trace.go:134]: Worth doing a length check on `vals`?

File: pkg/cli/debug_job_trace.go
Comment (pbardea) [inline pkg/cli/debug_job_trace.go:151]: These indexes are off by 1 (same with the checks below I think)

**Review by abarganier — COMMENTED** [review]
Nice to see this coming together, :lgtm: although I'll leave the approval to Paul

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66914#-:-MdSmeOv7Ln6YsvrKa03:bq0k7yd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru)

___
*[pkg/cli/debug_job_trace.go, line 223 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66914#-MdSiPwm6zNOhEL17-w3:-MdSi-1R9ETHIPmBx3fr:bq5qe8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7cb71f163b4ec1115167e12aa440b565dc7c84d/pkg/cli/debug_job_trace.go#L223)):*
> ```
> if err = rows.Next(vals); err == io.EOF {
> ```
optional nit: thoughts on using this as the `for` loop condition, performing flushAndReset/error checking once we exit the loop? Same behavior, but perhaps a bit safer against future maintainers accidentally introducing an infinite loop

File: pkg/cli/debug_job_trace.go
Comment (adityamaru) [reply pkg/cli/debug_job_trace.go:126]: Done.

File: pkg/cli/debug_job_trace.go
Comment (adityamaru) [reply pkg/cli/debug_job_trace.go:112]: Hmmm we do need to unmarshal the curTraceCollection before we can assign cumulativeTraceCollection to it. In other words, we're not doing any extra work between here and the beginning of the method. Unless your suggestion was a code restructure?

File: pkg/cli/debug_job_trace_test.go
Comment (adityamaru) [reply pkg/cli/debug_job_trace_test.go]: Copy pasta from explain_bundle_test, fixed.

File: pkg/cli/debug_job_trace.go
Comment (adityamaru) [reply pkg/cli/debug_job_trace.go]: done.

File: pkg/cli/debug_job_trace.go
Comment (adityamaru) [reply pkg/cli/debug_job_trace.go:134]: done.

File: pkg/cli/debug_job_trace.go
Comment (adityamaru) [reply pkg/cli/debug_job_trace.go]: yup, fixed.

File: pkg/cli/debug_job_trace.go
Comment (pbardea) [reply pkg/cli/debug_job_trace.go:112]: Ah, nevermind. Just got myself confused.

**Review by pbardea — APPROVED** [review]

**Comment by adityamaru** [comment]
> Nice to see this coming together, :lgtm: although I'll leave the approval to Paul
> 
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66914#-:-MdSmeOv7Ln6YsvrKa03:bq0k7yd)_ status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru)
> 
> _[pkg/cli/debug_job_trace.go, line 223 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66914#-MdSiPwm6zNOhEL17-w3:-MdSi-1R9ETHIPmBx3fr:bq5qe8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7cb71f163b4ec1115167e12aa440b565dc7c84d/pkg/cli/debug_job_trace.go#L223)):_
> 
> > ```
> > if err = rows.Next(vals); err == io.EOF {
> > ```
> 
> optional nit: thoughts on using this as the `for` loop condition, performing flushAndReset/error checking once we exit the loop? Same behavior, but perhaps a bit safer against future maintainers accidentally introducing an infinite loop

I tried it out but I _think_ this is a little easier to reason about (we also have this pattern in cli doctor and a couple of other places). The for loop header ends up looking like this:
```
var err error
for err = rows.Next(); err != io.EOF; err = rows.Next() {
}
if err == io.EOF {
flushAndReset()
}
```

My main concern is that the scoping of err might be easy to mess up. An infinite loop should not be possible since if we try to access a row once EOF has been returned, we'll get an exception.

**Comment by adityamaru** [comment]
TFTR!

bors r=pbardea,abarganier

**Pull request merged**
